### PR TITLE
Refactor the honest scripts

### PR DESCRIPTION
**Pull request by MedicineYeh**

### CODE REVIEW

File: libexec/honest-common.sh [hunk 1]
```diff
@@ -0,0 +1,124 @@
+###
```
Comment (david942j) [inline libexec/honest-common.sh]: Move this file to share/honest/

File: libexec/honest-clone [hunk 2]
```diff
@@ -1,83 +1,100 @@
...
-  return 0
+  local tag="$2"
+  local path="$3"
+  set -e
+  # TODO Use shallow clone or other techniques to improve the speed
+  git clone "$url" "$path" >/dev/null
+  cd $path
+  # if $tag is not present, checkout to the latest release.
+  tag="${tag:-$(git describe --tags `git rev-list --tags --max-count=1 2>/dev/null` 2>/dev/null)}"
+  [[ "$tag" != "" ]] && git checkout --detach "$tag" >/dev/null
```
Comment (david942j) [inline libexec/honest-clone]: I think it's ok to show the checking out message to users (no need >/dev/null)
Or maybe we can add silent/verbose options to control the logging level

File: libexec/honest-clone [hunk 3]
```diff
@@ -1,83 +1,100 @@
...
-      TAG=${url:$i+1}
-    fi
-  done
-  return 0
+  local tag="$2"
+  local path="$3"
+  set -e
+  # TODO Use shallow clone or other techniques to improve the speed
+  git clone "$url" "$path" >/dev/null
+  cd $path
```
Comment (david942j) [inline libexec/honest-clone]: Same as the comment at line 73

File: libexec/honest-common.sh [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  local index=0
+  shift
+  for e; do
+      [[ "$e" == "$match" ]] && echo $index && return 0;
+      index=$(( $index + 1 ))
+  done
+  echo "-1"
+}
+
+get_hashed_name() {
```
Comment (david942j) [inline libexec/honest-common.sh]: `get_tmp_dir`

**Comment by MedicineYeh** [comment]
>
/dev/null那是我延續你原本放在function那邊的，改成放到會有訊息的那兩行，這腳本因為是會被honest執行並且取回字串，所以兩邊還是要選一邊做啦，我當初改的時候還不是很確定要怎樣只是先把每項事物的影響層級盡量壓低，不要有在那一行看不出來的行為發生，例如我現在這樣放，平常debug時在function裡面echo就會有訊息印出來，不會不知道為什麼的沒訊息（原本做法是把redirection接在function
call）
不過說真的這還沒很確定要怎樣用，因為像是error
message也要能關靜音或者改用stderr不然honest呼叫honest-clone時有任何問題發生還是會導致回傳回來得字串除了path之外還有error
message，這樣rm -fr就會掛掉。

這部分是想說等跟你討論，你覺得呢？

**Comment by david942j** [comment]
Let me take a look

**Comment by MedicineYeh** [comment]
Your suggestions are adopted. Have a look.

**Comment by david942j** [comment]
And please fix the failed tests as well

**Comment by david942j** [comment]
* For tests failed becuz of error message being changed, feel free to change the test scripts
* But running `bin/honest github:david942j/honest gem:meowaaa` gets: `[ERROR] Fail on running 'honest-gem`.
A clearer error msg would be better.
(such as `Could not find a valid gem 'meowaaa' (>= 0) in any repository`)

**Comment by david942j** [comment]
by running `bin/honest github:david942j/honest gem:meowaaa`,
`[ERROR] Fail on running 'honest-gem` is output
I think a nice error msg would be better.
(such as `Could not find a valid gem 'meowaaa' (>= 0) in any repository`)

**Comment by david942j** [comment]
> /dev/null那是我延續你原本放在function那邊的，改成放到會有訊息的那兩行，這腳本因為是會被honest執行並且取回字串，所以兩邊還是要選一邊做啦，我當初改的時候還不是很確定要怎樣只是先把每項事物的影響層級盡量壓低，不要有在那一行看不出來的行為發生，例如我現在這樣放，平常debug時在function裡面echo就會有訊息印出來，不會不知道為什麼的沒訊息（原本做法是把redirection接在function
call）
不過說真的這還沒很確定要怎樣用，因為像是error
message也要能關靜音或者改用stderr不然honest呼叫honest-clone時有任何問題發生還是會導致回傳回來得字串除了path之外還有error
message，這樣rm -fr就會掛掉。
這部分是想說等跟你討論，你覺得呢？

How about let all logging functions `info/error` and `debug/warn (maybe in the future)` output to stderr?
And use an integral env (e.g. `HONEST_LOGGING_LEVEL`) to control if the message should be printed

**Comment by MedicineYeh** [comment]
All passed. Also, the log level feature is added and all the __log messages__ are now printed to stderr instead of stdout.

File: share/honest/honest-common.sh [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+HONEST_VERSION="0.0.0"
+
+# Currently supported package vendors
+VENDORS=("gem")
+REPO_VENDORS=("github" "bitbucket" "gitlab")
+REPO_VENDORS_HOST=("github.com" "bitbucket.org" "gitlab.com")
+
+# Set log level to stderr if not set. The number can be 0(any) - 5,
+# representing - debug, info, warn, error, and fatal
+HONEST_LOGGING_LEVEL=${HONEST_LOGGING_LEVEL:-0}
```
Comment (david942j) [inline share/honest/honest-common.sh]: set default value to 1?
(let the default be no debug message)

File: share/honest/honest-common.sh [hunk 6]
```diff
@@ -0,0 +1,177 @@
+###
+HONEST_VERSION="0.0.0"
+
+# Currently supported package vendors
+VENDORS=("gem")
+REPO_VENDORS=("github" "bitbucket" "gitlab")
+REPO_VENDORS_HOST=("github.com" "bitbucket.org" "gitlab.com")
+
+# Set log level to stderr if not set. The number can be 0(any) - 5,
```
Comment (david942j) [inline share/honest/honest-common.sh]: `Set log level to stderr if not set`
-> `Set log level to 1 if not set`

**Comment by david942j** [comment]
tests broken QQ
Set `TMPDIR` in test_helper.bash is a bad idea

File: libexec/honest [hunk 7]
```diff
@@ -85,129 +45,124 @@ push_tmp() {
...
+      ;;
+  esac
   if [[ $? -eq 0 ]]; then
     info "OK, ${YELLOW}$pkg${NC} is ${GREEN}Honest${NC}!"
     clear_tmp_dirs
   else
-    if [[ -t 1 ]]; then
-      (set -x; git diff --no-index --diff-filter=M "$REPO_PATH" "$PACKAGE_PATH"data)
-    fi
+    (git --no-pager diff --no-index --diff-filter=M "$REPO_PATH" "$PACKAGE_PATH"/data)
```
Comment (david942j) [inline libexec/honest]: `()` can be removed?

File: libexec/honest-clone [hunk 8]
```diff
@@ -1,83 +1,136 @@
 #!/usr/bin/env bash
 #
-# Usage: honest-clone <git-url>
-
+# Usage: honest-clone <git-url> [--quiet]
+#
+# Options:
+#        --quiet
+#               Hide the all the output messages except for the name of temp directory
```
Comment (david942j) [inline libexec/honest-clone]: Hide all the output messages except the name of temp directory

File: libexec/honest-clone [hunk 9]
```diff
@@ -1,83 +1,136 @@
...
+[[ "$OSTYPE" == "darwin"* ]] && readlink() { echo $(cd $(dirname "$2") && pwd)/$(basename "$2"); }
+
+SCRIPT_DIR="$(readlink -f "$(dirname "$0")")"
+QUITE_MODE="0"
+# Import common functions and variables of honest
+source "$SCRIPT_DIR"/../share/honest/honest-common.sh
+
+#######################################
+# Convert <vendor>:<author>/<project>[@..] to https://<vendor's url>/<author>/<project>[@..]
+# When failed to convert the format, the original string will be return.
```
Comment (david942j) [inline libexec/honest-clone]: be returned

File: libexec/honest-clone [hunk 10]
```diff
@@ -1,83 +1,136 @@
...
+  return 0
 }
 
-is_start_with() {
-  [[ "$1" != "$2"* ]] && return 1
-  local str="$1"
-  RESULT="${str:${#2}}"
+#######################################
+# Parse the tag string from git URL string which can be
+# 'https://<author>@github.com/<authors>/<repo>@v1.00'.
```
Comment (david942j) [inline libexec/honest-clone]: use `@v1.0.0` as example would be better

Comment (david942j) [inline libexec/honest-clone]: `<authors>/<repo>` -> `<author>/<repo>`

File: libexec/honest-clone [hunk 11]
```diff
@@ -1,83 +1,136 @@
...
+# The tag string will be 'v1.00' in this case.
+# Globals:
+#   None
+# Arguments:
+#   <Repo> The repo name with format '<vendor>:<author>/<project>[@..]'
```
Comment (david942j) [inline libexec/honest-clone]: The format should be `http(s)://URL[@..]`

File: libexec/honest-clone [hunk 12]
```diff
@@ -1,83 +1,136 @@
...
-die() {
-  local script_name=`basename "$0"`
-  echo "$script_name: $1"
-  exit 1
-}
+# Compatibility to MAC
+[[ "$OSTYPE" == "darwin"* ]] && readlink() { echo $(cd $(dirname "$2") && pwd)/$(basename "$2"); }
+
+SCRIPT_DIR="$(readlink -f "$(dirname "$0")")"
+QUITE_MODE="0"
```
Comment (david942j) [inline libexec/honest-clone]: Four typos in this file: `quite` -> `quiet`

**Review by david942j — CHANGES_REQUESTED** [review]

**Comment by MedicineYeh** [comment]
All fixed

File: libexec/honest-clone [hunk 13]
```diff
@@ -50,12 +50,12 @@ abbr_form() {
 
 #######################################
 # Parse the tag string from git URL string which can be
-# 'https://<author>@github.com/<authors>/<repo>@v1.00'.
+# 'https://<author>@github.com/<author>/<repo>@v1.0.0'.
 # The tag string will be 'v1.00' in this case.
```
Comment (david942j) [inline libexec/honest-clone]: 1.0.0 XD

Comment (david942j) [reply libexec/honest-clone]: but consider the latest pr.. alright, forget this

**Review by david942j — APPROVED** [review]

**Pull request merged**